ml/trainers.py

```diff
@@ -157,9 +157,9 @@ def train(self, minibatch=None, tol=None, maxepochs=None, monitor_every=None, pa
                 # print info
                 if verbose:
                     if self.do_validation:
-                        print 'Epoch = {0:.2f}, train loss = {1}, validation loss = {2}'.format(epoch, trn_loss, val_loss)
+                        print('Epoch = {0:.2f}, train loss = {1}, validation loss = {2}'.format(epoch, trn_loss, val_loss))
                     else:
-                        print 'Epoch = {0:.2f}, train loss = {1}'.format(epoch, trn_loss)
+                        print('Epoch = {0:.2f}, train loss = {1}'.format(epoch, trn_loss))
 
             # check for convergence
             if abs(diff) < tol or iter >= maxiter or patience_left <= 0:
```

Comment:
If you allow `tol` to be None, `abs(diff) < tol` will fail with a type error, as you can't compare a number with None. Please address this.

Reply:
From what I seemed to understand, If `tol` was `None`, it meant that the user did not want to use a tolerance criteria to stop the training process. Since comparing float to None is always false in python2, this was an effective method of achieving the latter.

That is why above I have replaced the default value of tol to `-1.0` (instead of `None`) and changed the corresponding assert. So when you don't want tolerance criteria to stop the training process, instead of using the value `None`, it uses `-1.0`. Since the value of tol is never changed again in this function the value of tol can no longer be `None`.

This is why I therefore believe that we never allow `tol` to be `None`.

However, I agree with you that it is perhaps more intuitive to use `None` instead of `-1.0` to indicate that the tolerance criteria to stop the training process is not wanted.

We could have something like this:

````
if tol is not None and abs(diff) < tol or iter >= maxiter or patience_left <= 0:
````

Obviously changing the assert and the default value from before.

I'm so sorry for the very long message!

Reply:
Ah yes you're right – my mistake. I got confused and I thought that you changed -1 to None, but you actually changed None to -1. The way you did it is perfectly fine.